Framework/DataHandling/src/AlignAndFocusPowderSlim.cpp

```diff
@@ -293,41 +261,61 @@ template <typename Type> std::pair<Type, Type> parallel_minmax(const std::vector
   }
 }
 
-template <typename CountsType> class ProcessEventsTask {
+class ProcessEventsTask {
 public:
-  ProcessEventsTask(const Histogrammer *histogrammer, const std::vector<detid_t> *detids,
-                    const std::vector<float> *tofs, const AlignAndFocusPowderSlim::BankCalibration *calibration,
-                    std::vector<CountsType> *y_temp, const std::set<detid_t> *masked)
-      : m_histogrammer(histogrammer), m_detids(detids), m_tofs(tofs), m_calibration(calibration), y_temp(y_temp),
-        masked(masked), no_mask(masked->empty()) {}
+  ProcessEventsTask(const std::vector<detid_t> *detids, const std::vector<float> *tofs,
+                    const AlignAndFocusPowderSlim::BankCalibration *calibration, const std::vector<double> *binedges,
+                    const std::set<detid_t> *masked)
+      : m_detids(detids), m_tofs(tofs), m_calibration(calibration), m_binedges(binedges), masked(masked),
+        no_mask(masked->empty()) {
+    y_temp.resize(m_binedges->size() - 1, 0);
+  }
+  // Local histogram for this block/thread
+  std::vector<uint32_t> y_temp;
 
-  void operator()(const tbb::blocked_range<size_t> &range) const {
-    // both iterators need to be incremented in each loop
-    auto detid_ptr = m_detids->cbegin() + range.begin();
-    auto tof_ptr = m_tofs->cbegin() + range.begin();
+  void operator()(const tbb::blocked_range<size_t> &range) {
+    // Fast histogramming without sorting or cumulative counting
+
+    // Precompute bin edges for fast lookup
+    const auto &binedges = *m_binedges;
+    const size_t nbins = binedges.size() - 1;
 
     for (size_t i = range.begin(); i < range.end(); ++i) {
-      if (no_mask || (!masked->contains(*detid_ptr))) {
-        // focussed time-off-flight
-        const auto tof = static_cast<double>(*tof_ptr) * m_calibration->value(*detid_ptr);
-        // increment the bin if it is found
-        if (m_histogrammer->inRange(tof)) {
-          if (const auto binnum = m_histogrammer->findBin(tof)) {
-            y_temp->at(binnum.value())++;
-          }
+      const detid_t detid = (*m_detids)[i];
+      if (no_mask || (masked->find(detid) == masked->end())) {
+        // Apply calibration
+        const double tof = static_cast<double>((*m_tofs)[i]) * m_calibration->value(detid);
+
+        // Find the bin index using binary search
+        const auto it = std::upper_bound(binedges.begin(), binedges.end(), tof);
```

Comment:
cbegin/cend lets the compiler do a little bit more for you
```suggestion
        const auto it = std::upper_bound(binedges.cbegin(), binedges.cend(), tof);
```